### DISCUSSION TITLE
How to limit max thrust for brushless / BigQuad build?

### DISCUSSION THREAD

**Discussion by schlagenhauf** [bitcraze/discussions#7] [Q&A]
Based on the BigQuad deck I have assembled a brushless quadcopter. It seems to work fine so far but I am hesitant to mount props because at full thrust the quad will be quite violent.

I'd like to reduce the maximum thrust / scale down the thrust mapping, but I'm unsure where. Reading through `motor.c` and `stabilizer.c` didn't yield any clues. Ideally the new limits would also be visible to the stabilizer to not create weird control behavior.

It seems bitcraze/crazyflie-firmware#365 is related?

Thanks!

**Comment by tobbeanton** [comment]
This is related to bitcraze/crazyflie-firmware#365 and still not implemented. You could do something temporary meanwhile but since I have not looked into it in detail I'm not 100% sure what the best solution is. Ultimately you would like to limit it at the driver level so it would be "impossible" to go over the limit. Looking at it quickly that is a bit complicated and I instead suggest to implement another limit function in [power_distribution_stock.c](https://github.com/bitcraze/crazyflie-firmware/blob/2a282b575c2541d3f3b4552296952a6cc40bf5b5/src/modules/src/power_distribution_stock.c#L54).

**Comment by schlagenhauf** [comment]
Thanks for your help!

Based on your hints i made this hack that seems to work (in that it limits the motor speed, no flight tests yet): https://github.com/schlagenhauf/crazyflie-firmware/tree/thrust_limit_hack.

Definitely not a proper solution, but i can keep my fingers. :partying_face:

**Reply by tobbeanton** [reply]
That is basically what I had in mind to :+1:

Showing a sniplet of your code:
```
uint16_t limitUint16toPercent(int32_t value, const uint16_t max_value_percent) {
  const uint16_t max_value = UINT16_MAX * max_value_percent / 100;
  uint16_t limited = limitUint16(value);
  if (limited > max_value) {
    limited = max_value;
  }
  return limited;
}

//#define limitThrust(VAL) limitUint16(VAL)
#define limitThrust(VAL) limitUint16ToPercent(VAL, 35)
```
